### PR TITLE
Package Project, Add Doc

### PR DESCRIPTION
**Pull request by LaurenceWarne**
Hi! I've made some changes which allow your project to be used as a library for use in other projects, along with a few other things:

- Add setup.py for easy installation and dependency management (you could also consider adding the package to the python package index?)
- Once the package is installed, you can still use `perlin2d.py` and `perlin3d.py` as scripts (`perlin-2d-noise` and `perlin-3d-noise` from the command line - assuming `~/.local/bin` is in your path)
- Update README with this installation info
- Add docstrings to main functions

### CODE REVIEW

File: perlin_numpy/perlin2d.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    return t*t*t*(t*(t*6 - 15) + 10)
+
+
+def generate_perlin_noise_2d(
+        shape, res, tileable=(False, False), interpolant=interpolant
+):
+    """Generate a 2D numpy array of perlin noise.
+
+    Args:
+        shape: The shape of the generated array (tupe of two ints).
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py]: typo "tupe"

File: perlin_numpy/perlin2d.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            (tuple of two bools). Defaults to (False, False).
+        interpolant: The, interpolation function, defaults to
+            t*t*t*(t*(t*6 - 15) + 10).
+
+    Returns:
+        A numpy array of fractal noise and of shape shape generated by
+        combining several octaves of perlin noise.
+
+    Raises:
+        ValueError: If res[0] does not divide shape[0] or res[1] does
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py]: The second part `lacunarity**(octaves-1)*res[0] does not divide shape[0] or lacunarity**(octaves-1)*res[1] does not divide shape[1]` is sufficient

File: perlin_numpy/perlin3d.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def generate_fractal_noise_3d(
+        shape, res, octaves=1, persistence=0.5, lacunarity=2,
+        tileable=(False, False, False), interpolant=interpolant
+):
+    """Generate a 3D numpy array of fractal noise.
+
+    Args:
+        shape: The shape of the generated array (tupe of three ints).
```
Comment (pvigier) [inline perlin_numpy/perlin3d.py]: typo "tupe"

File: perlin_numpy/perlin3d.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        interpolant: The, interpolation function, defaults to
+            t*t*t*(t*(t*6 - 15) + 10).
+
+    Returns:
+        A numpy array of fractal noise and of shape shape generated by
+        combining several octaves of perlin noise.
+
+    Raises:
+        ValueError: ValueError: If shape is not a multiple of res,
+            or if shape is not a mulitiple of
```
Comment (pvigier) [inline perlin_numpy/perlin3d.py]: typo "mulitiple"

File: perlin_numpy/perlin3d.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        tileable: If the noise should be tileable along each axis
+            (tuple of three bools). Defaults to (False, False, False)
+        interpolant: The interpolation function, defaults to
+            t*t*t*(t*(t*6 - 15) + 10).
+
+    Returns:
+        A numpy array of shape shape with the generated noise.
+
+    Raises:
+        ValueError: If shape is not a multiple of res.
```
Comment (pvigier) [inline perlin_numpy/perlin3d.py:27]: I prefer this format for the condition than the one used in perlin2d.py. It is more compact.

File: perlin_numpy/perlin3d.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            (tuple of three bools). Defaults to (False, False, False).
+        interpolant: The, interpolation function, defaults to
+            t*t*t*(t*(t*6 - 15) + 10).
+
+    Returns:
+        A numpy array of fractal noise and of shape shape generated by
+        combining several octaves of perlin noise.
+
+    Raises:
+        ValueError: ValueError: If shape is not a multiple of res,
```
Comment (pvigier) [inline perlin_numpy/perlin3d.py]: The first condition is not necessary.

**Review by pvigier — COMMENTED** [review]
Hi!

Thanks for the PR.

I let some inline comments.

In addition, I wonder why there are as many requirements in requirements.txt. I think numpy is the only requirement for the functions and matplotlib for the examples.

By the way, wouldn't it be better if the example snippets are moved elsewhere? Maybe in the README (this way we can get rid of the requirements of matplotlib) ?

Finally, I noticed that you reformat some pieces of code, do you follow a specific style convention or use an autoformat tool? I am not fond of the lines with only a parenthesis and the use of spaces when indexing arrays is not uniform in the whole code (I thought it was more common to not put spaces after ',' when indexing multidimensional arrays with numpy). I am not against a change in the coding style, just so that we agree on one.

Best,
Pierre

**Comment by LaurenceWarne** [comment]
Hi, thanks for getting back to me.

Apologies for the eye-sore typos, I think I understand all of your comments, I'll get on them (:

Regarding the dependencies, matplotlib has a lot of dependencies [itself](https://matplotlib.org/3.1.1/users/installing.html#dependencies), so yeah with respect to your second point, dropping the dependency would remove the need for a lot of the existing dependencies. I can do this, or perhaps this would be better discussed/implemented in a different PR?

I have flake8 integrated into my IDE which displays warnings if code layout goes against the PEP guidelines, for [code layout](https://www.python.org/dev/peps/pep-0008/#id16), there are a few acceptable ways. For example I could change snippets like:

```
foo = long_function_name(
    var_one, var_two,
    var_three, var_four
)
```
to:

```
foo = long_function_name(var_one, var_two,
                         var_three, var_four)
```

If this is preferable? Yes I looked and the numpy docs appear to follow this convention, I'll revert these changes. Though there some existing lines did not follow this, e.g.

```python
g00 = gradients[    :-d[0],    :-d[1]]
```

Of course I can keep these the same if want.

Cheers ~ Laurence

**Comment by pvigier** [comment]
> Regarding the dependencies, matplotlib has a lot of dependencies itself, so yeah with respect to your second point, dropping the dependency would remove the need for a lot of the existing dependencies.

Is it necessary to list transitive dependencies in requirements.txt or just putting matplotlib is sufficient? Moreover, there are strong equalities `==` for versions in requirements.txt, shouldn't we use `>=`? 

>  I can do this, or perhaps this would be better discussed/implemented in a different PR?

I have nothing against doing it in this PR as it reorganizes the project to make it "packageable". But if you prefer, I can do it later in another PR.

> I have flake8 integrated into my IDE which displays warnings if code layout goes against the PEP guidelines, for code layout, there are a few acceptable ways.

OK! Choose the one you prefer.

> Though there some existing lines did not follow this

Yes, please keep them, I add extra spaces to align the indices of each dimension for readability and to make the code more understandable.

**Comment by LaurenceWarne** [comment]
Thanks for your fast reply,

> Is it necessary to list transitive dependencies in requirements.txt or just putting matplotlib is sufficient? Moreover, there are strong equalities == for versions in requirements.txt, shouldn't we use >=?

In terms of the requirements.txt file, the way I created it was just installing all of the dependencies on a new virtual environment and then just calling:
```
pip3 freeze > requirements.txt
```
So that the process is automated. Then I can install all the dependencies by just calling:

```
pip3 install -r requirements.txt
```

So the requirements.txt often lists transitive dependents in a requirements.txt since most of the time its made by a ```pip freeze``` which does list transitive dependencies. But numpy has no transitive dependencies, once we uninstall matplotlib, this shouldn't be a problem. 

In terms of a numpy version to support, I think the project will run on any python 3 installation, so according to https://numpy.org/neps/nep-0029-deprecation_policy.html any version numpy version >= 1.15 will be ok. I'll change the requirements.txt to this effect.

> I have nothing against doing it in this PR as it reorganizes the project to make it "packageable". But if you prefer, I can do it later in another PR.

No worries, I can do it. I think a good way to do this would be to change the "gallery" section in the README to a "recipes" section, so that we can couple the code snippets along with the images they generate. What do you think?

> Yes, please keep them, I add extra spaces to align the indices of each dimension for readability and to make the code more understandable.

Ok, no problem.

**Comment by pvigier** [comment]
> No worries, I can do it. I think a good way to do this would be to change the "gallery" section in the README to a "recipes" section, so that we can couple the code snippets along with the images they generate. What do you think?

Good idea! Let me know if you need help.

**Comment by LaurenceWarne** [comment]
Ok! I think I've addressed everything now, what do you think? (:

File: README.md [hunk 7]
```diff
@@ -49,9 +66,85 @@ The function `generate_fractal_noise_2d` combines several octaves of 3D perlin n
 
 Note: `shape` must be a multiple of `lacunarity^(octaves-1)*res`
 
-## Gallery
+## Recipes
+
+Note these snippets require ![matplotlib](https://matplotlib.org/).
```
Comment (pvigier) [inline README.md]:
```suggestion
Note these snippets require [matplotlib](https://matplotlib.org/).
```

File: perlin_numpy/perlin2d.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+def generate_perlin_noise_2d(
+        shape, res, tileable=(False, False), interpolant=interpolant
+):
+    """Generate a 2D numpy array of perlin noise.
+
+    Args:
+        shape: The shape of the generated array (tuple of two ints).
+        res: The number of periods of noise to generate along each
+            axis (tuple of two ints). Note res[0] must divide shape[0]
+            and res[1] must divide shape[1].
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py]:
```suggestion
            axis (tuple of two ints). Note shape must be a multiple of
            res.
```

File: perlin_numpy/perlin2d.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        interpolant=interpolant
+):
+    """Generate a 2D numpy array of fractal noise.
+
+    Args:
+        shape: The shape of the generated array (tuple of two ints).
+            This must be a multiple of lacunarity**(octaves-1)*res.
+        res: The number of periods of noise to generate along each
+            axis (tuple of two ints). Note res[0] must divide shape[0]
+            and res[1] must divide shape[1].
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py]:
```suggestion
            axis (tuple of two ints). Note shape must be a multiple of
            (lacunarity**(octaves-1)*res).
```

Or maybe remove the note, as there is already a note in the description of `shape`.

File: perlin_numpy/perlin3d.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        tileable=(False, False, False), interpolant=interpolant
+):
+    """Generate a 3D numpy array of fractal noise.
+
+    Args:
+        shape: The shape of the generated array (tuple of three ints).
+            This must be a multiple of lacunarity**(octaves-1)*res.
+        res: The number of periods of noise to generate along each
+            axis (tuple of three ints). Note shape must be a multiple of
+            res.
```
Comment (pvigier) [inline perlin_numpy/perlin3d.py]:
```suggestion
            axis (tuple of three ints). Note shape must be a multiple of
            (lacunarity**(octaves-1)*res).
```

File: perlin_numpy/perlin3d.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+
+def generate_perlin_noise_3d(
+        shape, res, tileable=(False, False, False),
+        interpolant=interpolant
+):
+    """Generate a 3D numpy array of perlin noise.
+
+    Args:
+        shape: The shape of the generated array (tuple of three ints).
```
Comment (pvigier) [inline perlin_numpy/perlin3d.py:13]:
```suggestion
        shape: The shape of the generated array (tuple of three ints).
            This must be a multiple of res.
```

File: perlin_numpy/perlin2d.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    return t*t*t*(t*(t*6 - 15) + 10)
+
+
+def generate_perlin_noise_2d(
+        shape, res, tileable=(False, False), interpolant=interpolant
+):
+    """Generate a 2D numpy array of perlin noise.
+
+    Args:
+        shape: The shape of the generated array (tuple of two ints).
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py:14]:
```suggestion
        shape: The shape of the generated array (tuple of two ints).
            This must be a multiple of res.
```

File: perlin_numpy/perlin2d.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        interpolant: The, interpolation function, defaults to
+            t*t*t*(t*(t*6 - 15) + 10).
+
+    Returns:
+        A numpy array of fractal noise and of shape shape generated by
+        combining several octaves of perlin noise.
+
+    Raises:
+        ValueError: If (lacunarity**(octaves-1)*res) does not divide
+            shape.
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py]:
```suggestion
        ValueError: If shape is not a multiple of
            (lacunarity**(octaves-1)*res).
```

File: perlin_numpy/perlin2d.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        tileable: If the noise should be tileable along each axis
+            (tuple of two bools). Defaults to (False, False)
+        interpolant: The interpolation function, defaults to
+            t*t*t*(t*(t*6 - 15) + 10).
+
+    Returns:
+        A numpy array of shape shape with the generated noise.
+
+    Raises:
+        ValueError: If res does not divide shape.
```
Comment (pvigier) [inline perlin_numpy/perlin2d.py]:
```suggestion
        ValueError: If shape is not a multiple of res.
```

File: requirements.txt [hunk 15]
```diff
@@ -0,0 +1 @@
+-e .
```
Comment (pvigier) [inline requirements.txt:1]: Can you explain me what this produce?

**Comment by pvigier** [comment]
I am nitpicking on some minor points to make the doc uniform. I can do the commit to fix these if you want.

Otherwise, it looks good to me! I will merge the PR after the changes.

File: requirements.txt (hunk 15 above)
Comment (LaurenceWarne) [reply requirements.txt:1]: If you run ```pip install -r requirements.txt```, putting this in the requirements.txt will run ```pip install -e .``` which will get the dependencies from the setup.py file. I think this is nicer than what was there before.

**Comment by LaurenceWarne** [comment]
Ok, the doc should all be consistent now! Yep, if I've introduced more typos into the doc it might be faster if you just make a quick commit through GH (: , but thanks for taking the time to review all the changes, I appreciate it.

**Comment by pvigier** [comment]
Thank you very much for the PR. I will see to register the library in PyPI. And add an AUTHORS file and a section in the README to credit you properly.

Have a nice day!

**Pull request merged**